amethyst_ui/src/text.rs

```diff
@@ -1,27 +1,42 @@
-use amethyst_assets::{AssetStorage, Loader};
-use amethyst_renderer::{Texture, TextureData, TextureHandle, TextureMetadata};
-use gfx::format::{ChannelType, SurfaceType};
-use rusttype::{Point, Scale};
-use specs::{Component, DenseVecStorage, Fetch, Join, ReadStorage, System, WriteStorage};
+use std::cmp::Ordering;
+use std::ops::Range;
+
+use amethyst_core::timing::Time;
+use clipboard::{ClipboardContext, ClipboardProvider};
+use hibitset::BitSet;
+use rusttype::PositionedGlyph;
+use shrev::{EventChannel, ReaderId};
+use specs::{Component, DenseVecStorage, Entities, Entity, Fetch, FetchMut, Join, ReadStorage,
+            System, WriteStorage};
 use unicode_normalization::UnicodeNormalization;
 use unicode_normalization::char::is_combining_mark;
+use unicode_segmentation::UnicodeSegmentation;
+use winit::{ElementState, Event, KeyboardInput, ModifiersState, MouseButton, VirtualKeyCode,
+            WindowEvent};
 
 use super::*;
 
-/// A component used to display text in this entities UiTransform
+/// A component used to display text in this entity's UiTransform
+#[derive(Clone, Derivative)]
+#[derivative(Debug)]
 pub struct UiText {
-    /// The texture that text is rendered onto.  None if text isn't rendered yet.
-    pub(crate) texture: Option<TextureHandle>,
-    /// The font used to display the text.
-    font: FontHandle, //TODO: Mark this component dirty when the font is reloaded.
-    /// The text being displayed
-    text: String,
-    /// The normalized RGBA color of the text being displayed
-    color: [f32; 4],
-    /// The font size of the text being displayed
-    font_size: f32,
-    /// This is true if the texture needs to be re-rendered
-    dirty: bool,
+    /// The string rendered by this.
+    pub text: String,
+    /// The height of a line of text in pixels.
+    pub font_size: f32,
+    /// The color of the rendered text, using a range of 0.0 to 1.0 per channel.
+    pub color: [f32; 4],
+    /// The font used for rendering.
+    pub font: FontHandle,
+    /// If true this will be rendered as dots instead of the text.
+    pub password: bool,
```

Comment:
Not sure `password` is the correct field name for this. `hidden` maybe? Or maybe an enum for how to render the text (Normal, Hidden, Dotted) or some such ? Not sure, just an idea.

Reply:
What would be the difference between Hidden and Dotted?

Reply:
Hidden = nothing, Dotted= the current behavior for password

Reply:
What would be the use case for hidden in this context? I know sometimes that's used for password entry in console programs but I haven't seen anywhere else use it.